### PR TITLE
[bluestore][tools] Inspect allocations in bluestore

### PR DESCRIPTION
**Pull request by aclamk**
Toolkit to inspect in detail allocations that BlueStore makes. 
Allows to inspect block device allocations and all BlueFS used devices.
Provided in two lines:
1) ceph-bluestore-tool new operations fragmentation-dump and fragmentation-score
`./bin/ceph-bluestore-tool --path dev/osd0/ --allocator block fragmentation-score`
2) admin-socket interface
`./bin/ceph --admin-daemon /tmp/ceph-asok.*/osd.0.asok bluestore allocator score bluefs-db`

Fixes: https://tracker.ceph.com/issues/40704

### CODE REVIEW

File: src/os/bluestore/Allocator.cc [hunk 1]
```diff
@@ -26,13 +26,23 @@ class Allocator::SocketHook : public AdminSocketHook {
                                            "dump allocator free regions");
       if (r != 0)
         alloc = nullptr; //some collision, disable
+      if (alloc) {
+        r = admin_socket->register_command(("bluestore allocator score " + name).c_str(),
+                                           ("bluestore allocator score " + name).c_str(),
+                                           this,
+                                           "give measurement on allocator fragmentation quality (0-bad, 1-perfect)");
```
Comment (jdurgin) [inline src/os/bluestore/Allocator.cc]: does 'perfect' mean entirely fragmented? on an empty cluster this score is very close to 0

Comment (aclamk) [reply src/os/bluestore/Allocator.cc]: @jdurgin Text was incorrect. Now fixed I hope. Please check.

File: src/os/bluestore/BlueStore.h [hunk 2]
```diff
@@ -2209,6 +2210,9 @@ class BlueStore : public ObjectStore,
   */
   int _open_db_and_around(bool read_only);
   void _close_db_and_around();
+  int cold_open();
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h]: make it public and remove the trick with friend class?

**Comment by ifed01** [comment]
@aclamk - and one general complaint - IMO it's better to have commit as complete and minimal as possible. So it's better to avoid subsequent fixes or refactoring related to the previous commits from this PR. As well as multiple unrelated changes within a single commit.
It's better for both reviewing (if it's done on per-commit basis) and backporting when we might need just a part of the functionality.

File: src/os/bluestore/bluestore_tool.cc [hunk 3]
```diff
@@ -239,10 +252,26 @@ int main(int argc, char **argv)
...
+        "bluefs-bdev-expand, "
+        "bluefs-bdev-new-db, "
+        "bluefs-bdev-new-wal, "
+        "bluefs-bdev-migrate, "
+        "show-label, "
+        "set-label-key, "
+        "rm-label-key, "
+        "prime-osd-dir, "
+        "bluefs-log-dump, "
+        "fragmentation-dump, "
```
[the remark this answers is not in the document]
Comment (aclamk) [reply src/os/bluestore/bluestore_tool.cc]: Yes, "free-dump" is much better.

File: src/os/bluestore/BlueStore.h (hunk 2 above)
Comment (aclamk) [reply src/os/bluestore/BlueStore.h]: Done.

File: src/os/bluestore/fastbmap_allocator_impl.h [hunk 4]
```diff
@@ -141,6 +142,7 @@ class AllocatorLevel01Loose : public AllocatorLevel01
     L1_ENTRY_NOT_USED = 0x02,
     L1_ENTRY_FREE = 0x03,
     CHILD_PER_SLOT = bits_per_slot / L1_ENTRY_WIDTH, // 32
+    L1_ENTRIES_PER_SLOT = bits_per_slot / L1_ENTRY_WIDTH, //32
```
[the remark this answers is not in the document]
Comment (aclamk) [reply src/os/bluestore/fastbmap_allocator_impl.h]: I think these names better explain role of constant.
I made alias as I didn't want to change existing code.

Maybe if I explicitly assign `L1_ENTRIES_PER_SLOT = CHILD_PER_SLOT` ?

**Comment by aclamk** [comment]
jenkins make check

**Comment by aclamk** [comment]
jenkins retest this please

File: src/os/bluestore/fastbmap_allocator_impl.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/fastbmap_allocator_impl.h]: > I think these names better explain role of constant.
> I made alias as I didn't want to change existing code.
> 
> Maybe if I explicitly assign `L1_ENTRIES_PER_SLOT = CHILD_PER_SLOT` ?

Suggest to use existing names in this PR and made a standalone PR to do the rename

File: src/os/bluestore/BlueFS.cc [hunk 5]
```diff
@@ -42,6 +43,78 @@ static void slow_discard_cb(void *priv, void* priv2) {
...
+      stringstream ss;
+      bool r = true;
+      if (command == "bluestore bluefs available") {
+        int64_t alloc_size = 0;
+        cmd_getval(bluefs->cct, cmdmap, "alloc_size", alloc_size);
+        if ((alloc_size & (alloc_size - 1)) != 0) {
+          ss << "Invalid allocation size:'" << alloc_size << std::endl;
+        }
+        if (alloc_size == 0)
+          alloc_size = bluefs->cct->_conf->bluefs_alloc_size;
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: please note that this is going to be changed with Neha's https://github.com/ceph/ceph/pull/29404

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -5619,6 +5619,25 @@ int BlueStore::allocate_bluefs_freespace(
...
+    size_t dist_to_alignment;
+    size_t offset_in_block = off & (alloc_size - 1);
+    if (offset_in_block == 0)
+      dist_to_alignment = 0;
+    else
+      dist_to_alignment = alloc_size - offset_in_block;
+    if (dist_to_alignment >= len)
+      return;
+    len -= dist_to_alignment;
+    total += len & ~(alloc_size - 1);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: you might want to use p2align macro to round value down

File: src/os/bluestore/BlueStore.cc [hunk 7]
```diff
@@ -5619,6 +5619,25 @@ int BlueStore::allocate_bluefs_freespace(
   return 0;
 }
 
+size_t BlueStore::available_freespace(uint64_t alloc_size) {
```
Comment (jdurgin) [inline src/os/bluestore/BlueStore.cc]: is there any locking required here?

Comment (aclamk) [reply src/os/bluestore/BlueStore.cc]: Locking is done in method `dump` of Allocator.

File: src/os/bluestore/BlueStore.cc (hunk 6 above)
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc]: Done.

**Comment by aclamk** [comment]
Jenking, please check docs

**Comment by neha-ojha** [comment]
rados run: http://pulpito.ceph.com/nojha-2019-08-07_00:01:13-rados-wip-bluestore-monitor-allocations-distro-basic-smithi/ - failures are unrelated

**Comment by mnaser** [comment]
Did I miss something or the health checks document was updated without the actual health checks being added?

**Pull request merged**